taichi/ir/frontend_ir.cpp

```diff
@@ -721,28 +741,98 @@ bool IndexExpression::is_global() const {
   return is_field() || is_matrix_field() || is_ndarray();
 }
 
+static void field_validation(FieldExpression *field_expr,
+                             int index_dim,
+                             bool has_slice) {
+  TI_ASSERT(field_expr != nullptr);
+  if (field_expr->snode == nullptr)
+    return;
+  int field_dim = field_expr->snode->num_active_indices;
+
+  int has_dynamic = 0;
+  if (field_expr->snode->parent &&
+      field_expr->snode->parent->type == SNodeType::dynamic) {
+    has_dynamic = 1;
+  }
+  /*
+    Indexing to dynamic snode can be special, you can either index all the way
```

Comment:
`dynamic` SNode is already handled in a special way: https://github.com/taichi-dev/taichi/blob/59d8909112a924125ca6a13ceaa05d2219a359c0/python/taichi/lang/ast/ast_transformer.py#L800-L819

So maybe we don't need to handle this again here? BTW we need to raise `TaichiIndexError` instead of `TaichiTypeError`.

Reply:
Personally I'd prefer to have that checker in `IndexExpression::type_check()` as well, as an ultimate validation for the legality of an `IndexExpression` in Frontend IR semantics. 

On the other hand, the checker at Python level can trigger an early stop to provide a more accurate error message, but it could get bypassed since this may not be the only interface to create an `IndexExpression`

Reply:
Ah, the main point here is that `a[i, j].append(x)` is currently a syntax sugar for `ti.append(a.parent(), (i, j), x)` in Python, and `a[i, j]` itself is not valid. So IMO we don't need to (or shouldn't?) allow it in frontend IR.

Reply:
I see, I can remove this check in IndexExpression. But honestly this "syntax sugar" allows the use of intermediate(non-place) SNode and does hacks IndexExpression. If other frontend languages (say Rust) doesn't have the same "syntax sugar" then there will be "Python-specifc hacks" at Frontend IR level.

Reply:
I agree. I guess the root problem here is that SNodes are not in the type system yet, so handling these intermediate SNodes always needs hacks somewhere. Once we have SNodes in our type system, we can precisely describe rules like `a[i, j]` is allowed to have `append` attribute.

Reply:
Btw, there's no `TaichiIndexError ` at C++ level yet. Let's comply to `TaichiSyntaxError` then?

<img width="586" alt="image" src="https://user-images.githubusercontent.com/22334008/209906420-4e8ea623-be68-4894-86c5-3803ab23544a.png">

Reply:
Hmm I mention `TaichiIndexError` because it seems that the dynamic SNode hack relies on it: https://github.com/taichi-dev/taichi/blob/f4e3aa41e050408974bc6dac9bebe1c3ebc57d5d/python/taichi/lang/ast/ast_transformer.py#L822